src/v/serde/parquet/shredder.cc

```diff
@@ -0,0 +1,309 @@
+    // additional repetitions need to use the current repetition_depth as the
+    // repetition_level.
+    uint8_t repetition_depth = 0;
+};
+
+class record_shredder {
+public:
+    record_shredder(
+      const schema_element& root,
+      group_value record,
+      absl::FunctionRef<ss::future<>(shredded_value)> cb)
+      : _callback(cb) {
+        if (root.repetition_type != field_repetition_type::required) {
+            throw std::runtime_error("schema root nodes must be required");
+        }
+        _queue.emplace_back(
+          &root, traversal_levels{}, value(std::move(record)));
+    }
+
+    ss::future<> shred() {
+        while (!_queue.empty()) {
+            auto [element, levels, value] = std::move(_queue.back());
+            _queue.pop_back();
+            if (element->is_leaf()) {
+                co_await process_leaf_value(element, levels, std::move(value));
+            } else {
+                co_await process_group_node(element, levels, std::move(value));
+            }
+        }
+    }
+
+private:
+    ss::future<> process_leaf_value(
+      const schema_element* element, traversal_levels levels, value val) {
+        return ss::visit(
+          std::move(val),
+          [this, element, levels](repeated_value& list) {
+              return process_repeated_leaf_value(
+                element, std::move(list), levels);
+          },
+          [element](group_value&) {
+              return ss::make_exception_future(std::runtime_error(fmt::format(
+                "unexpected struct value for leaf schema element {}",
+                element->name)));
+          },
+          [this, element, levels](null_value& v) {
+              return _callback({
+                .schema_element_position = element->position,
+                .val = v,
+                .rep_level = levels.repetition_level,
+                .def_level = levels.definition_level,
+              });
+          },
+          [this, element, levels](auto& v) {
+              traversal_levels leaf_levels = levels;
+              if (element->repetition_type != field_repetition_type::required) {
+                  ++leaf_levels.definition_level;
+              }
+              return _callback({
+                .schema_element_position = element->position,
+                .val = value(std::move(v)),
+                .rep_level = leaf_levels.repetition_level,
+                .def_level = leaf_levels.definition_level,
+              });
+          });
+    }
+
+    ss::future<> process_repeated_leaf_value(
+      const schema_element* element,
+      repeated_value list,
+      traversal_levels levels) {
+        // Empty lists are treated as `null`
+        if (list.empty()) {
+            co_return co_await _callback({
+              .schema_element_position = element->position,
+              .val = null_value(),
+              .rep_level = levels.repetition_level,
+              .def_level = levels.definition_level,
+            });
```

Comment:
Just an observation, it's tempting to think that we should be calling process_optional_null_value() here as we do for the non-leaves.

Reply:
I renamed to `process_optional_null_group` - does that help?

Reply:
It does! Thanks